crates/bevy_ui/src/render/material_pipeline.rs

```diff
@@ -0,0 +1,659 @@
+) {
+    ui_meta.vertices.clear();
+
+    extracted_uinodes
+        .uinodes
+        .sort_by_key(|node| node.stack_index);
+    for extracted_uinode in extracted_uinodes.uinodes.drain(..) {
+        let uinode_rect = extracted_uinode.rect;
+
+        let rect_size = uinode_rect.size().extend(1.0);
+
+        let positions = QUAD_VERTEX_POSITIONS
+            .map(|pos| (extracted_uinode.transform * (pos * rect_size).extend(1.)).xyz());
+
+        let positions_diff = if let Some(clip) = extracted_uinode.clip {
+            [
+                Vec2::new(
+                    f32::max(clip.min.x - positions[0].x, 0.),
+                    f32::max(clip.min.y - positions[0].y, 0.),
+                ),
+                Vec2::new(
+                    f32::min(clip.max.x - positions[1].x, 0.),
+                    f32::max(clip.min.y - positions[1].y, 0.),
+                ),
+                Vec2::new(
+                    f32::min(clip.max.x - positions[2].x, 0.),
+                    f32::min(clip.max.y - positions[2].y, 0.),
+                ),
+                Vec2::new(
+                    f32::max(clip.min.x - positions[3].x, 0.),
+                    f32::min(clip.max.y - positions[3].y, 0.),
+                ),
+            ]
+        } else {
+            [Vec2::ZERO; 4]
+        };
+
+        let positions_clipped = [
+            positions[0] + positions_diff[0].extend(0.),
+            positions[1] + positions_diff[1].extend(0.),
+            positions[2] + positions_diff[2].extend(0.),
+            positions[3] + positions_diff[3].extend(0.),
+        ];
+
+        let transformed_rect_size = extracted_uinode.transform.transform_vector3(rect_size);
+
+        if extracted_uinode.transform.x_axis[1] == 0.0
+            && positions_diff[0].x - positions_diff[1].x >= transformed_rect_size.x
+            || positions_diff[1].y - positions_diff[2].y >= transformed_rect_size.y
+        {
+            continue;
+        }
+        let extent = uinode_rect.max;
+        let uvs = [
+            Vec2::new(
+                uinode_rect.min.x + positions_diff[0].x,
+                uinode_rect.min.y + positions_diff[0].y,
+            ),
+            Vec2::new(
+                uinode_rect.max.x + positions_diff[1].x,
+                uinode_rect.min.y + positions_diff[1].y,
+            ),
+            Vec2::new(
+                uinode_rect.max.x + positions_diff[2].x,
+                uinode_rect.max.y + positions_diff[2].y,
+            ),
+            Vec2::new(
+                uinode_rect.min.x + positions_diff[3].x,
+                uinode_rect.max.y + positions_diff[3].y,
+            ),
+        ]
+        .map(|pos| pos / extent);
+
+        for i in QUAD_INDICES {
+            ui_meta.vertices.push(crate::UiVertex {
+                position: positions_clipped[i].into(),
+                uv: uvs[i].into(),
+                color: Color::WHITE.into(),
+                mode: 1,
```

Comment:
If `mode` is always going to be set to `0`, shouldn't we add a separate `UiMaterialVertex` with `mode` removed? The batching optimization it allows can't be implemented with this design anyway (and probably aren't necessary).

Reply:
I did just as you suggested, had to make some other changes as well 👍 

thanks for the suggestion!